client/test/role.go

```diff
@@ -122,15 +125,20 @@ func ExecuteTwoPartyTest(ctx context.Context, role [2]Executer, cfg ExecConfig)
 	for i := 0; i < len(role); i++ {
 		wg.Add(1)
 		go func(i int) {
-			defer wg.Done()
 			log.Infof("Executing role %d", i)
 			role[i].Execute(cfg)
+			wg.Done()
 		}(i)
 	}
 
 	// wait for clients to finish or timeout
-	wg.WaitCtx(ctx)
-	return ctx.Err()
+	select {
+	case <-wg.WaitCh():
+	case <-ctx.Done():
+		t.Fatal(ctx.Err())
```

Comment:
Another approach would be to call `runtime.GoExit()` inside of `RequireNoError()` etc. which will then end the current go-routine.

Reply:
Thx for checking. In fact, with the previous version, the same problem exists. So we don't solve this here but at least we can let the test quickly know that something is going wrong. To be honest, it also doesn't seem super problematic that we don't shutdown the go routines as this happens automatically when the test fails. In fact, it might be problematic to call `runtime.GoExit()` (I was considering that), because then you might accidentally shutdown the main routine if the tests are not setup as expected.

Reply:
Yes, I also saw that in principle the same problem existed before and is also the reason why using `require` does not work in go-routines since it also just calls `t.FailNow()`.

And true, using the approach with `runtime.GoExit()` may cause unwanted behavior when using it without caution.